pkg/cloudprovider/aws/fleet/packing/packing.go

```diff
@@ -45,12 +47,108 @@ func NewPacker(ec2 ec2iface.EC2API) *PodPacker {
 // Pack returns the packings for the provided pods. Computes a set of viable
 // instance types for each packing of pods. Instance variety enables EC2 to make
 // better cost and availability decisions.
+// Pods provided are all schedulable in the same zone as tightly as possible.
 func (p *PodPacker) Pack(ctx context.Context, pods []*v1.Pod) ([]*Packings, error) {
 	zap.S().Debugf("Successfully packed %d pods onto %d nodes", len(pods), 1)
-	return []*Packings{
-		{
-			InstanceTypeOptions: []string{"m5.large"}, // TODO, prioritize possible instance types
-			Pods:                pods,
-		},
-	}, nil
+
+	// 1. Arrange pods in decreasing order by the amount of CPU requested, if
+	// CPU requested is equal compare memory requested.
+	sort.Sort(sort.Reverse(byResourceRequested{pods}))
+
+	// 2. Get all available instance types with there capacity and cost
+	// TODO add filters
+	instanceTypes := p.getInstanceTypes("")
+	// TODO reserve (Kubelet + daemon sets) overhead for instance types
+	return p.packSortedPods(pods, instanceTypes)
+}
+
+// takes a list of pods sorted based on their resource requirements compared by CPU and memory.
+
+func (p *PodPacker) packSortedPods(pods []*v1.Pod, instanceTypes []*instanceType) ([]*Packings, error) {
+	// Start with the smallest instance type and the biggest pod check how
+	// many pods can we fit, go to the next bigger type and check if we can fit
+	// more pods. Compare pods packed on all and select the instance type with
+	// highest pod count.
+	podsPacked := map[*v1.Pod]bool{}
+	packings := make([]*Packings, 0)
```

Comment:
Is there a reason for the `make(..., 0)` style here?

Reply:
Just bad habit, I tried to fix in all the places I have used it.